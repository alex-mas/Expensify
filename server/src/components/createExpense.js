import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './expenseForm';
import { addExpense } from '../actions/expenses';


export class CreateExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        //redirect user back to base route
        this.props.history.push('/dashboard');
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div> 
                </div>
                <div className="content-container">
                    <ExpenseForm onSubmit={this.onSubmit} mode={'Add'} />
                </div>
            </div>
        );
    }
}


const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (expense) => dispatch(addExpense(expense))
    }
}
export default connect(undefined, mapDispatchToProps)(CreateExpensePage);