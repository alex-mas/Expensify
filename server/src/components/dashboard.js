import React from 'react';
import ExpenseList from "./expenseList";
import ExpenseListFilters from './expenseListFilters';
import ExpensesSummary from './expensesSummary';
import ExpensesChart from './expensesChart';

const Dashboard = (props) => {
    return (
        <div>
            <ExpensesSummary />
            <ExpenseListFilters />
            <ExpenseList />
            <ExpensesChart/>
        </div>
    );
}

export default Dashboard;