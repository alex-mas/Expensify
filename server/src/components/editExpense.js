import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './expenseForm';
import { editExpense } from '../actions/expenses';



export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/dashboard');
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        history={this.props.history}
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                        mode={'Edit'}
                    />
                </div>
            </div>
        );
    }
}



const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
})

const bootstraper = function (state, props) {
    return {
        //returns the expense if both id matches
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    }
}
export default connect(bootstraper, mapDispatchToProps)(EditExpensePage);