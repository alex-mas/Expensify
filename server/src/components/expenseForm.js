import React from 'react';
import moment from 'moment';
import {connect} from 'react-redux';
import { SingleDatePicker } from 'react-dates';
import { removeExpense } from '../actions/expenses';
import 'react-dates/lib/css/_datepicker.css';




//Form to create expenses + input validation/sanitization
export class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);
        const exists = Boolean(props.expense);
        this.state = {
            description: exists ? props.expense.description : '',
            note: exists ? props.expense.note : '',
            //format the property so it matches regexp requirements
            amount: exists ? (props.expense.amount / 100).toString() : '',
            createdAt: exists ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            submitError: '',
            mode: props.mode ? props.mode : 'create'
        };
    }

    /* Event handlers*/
    //listining usually on onChange/onSubmit={methodname} or other change methods provided by third party components
    //for example: onDateChange={methodname} in the case of SingleDatePicker
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        //check that the value is either an empty string or a number that can have up to 2 decimals via regexp
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = (createdAt) => {
        //check that user hasn't deleted the date
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }

    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        //check for missing properties
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({
                submitError: 'Please provide amount and description before submiting'
            }));
        } else {
            //clear error message as the provided props are correct
            this.setState(() => ({
                submitError: ''
            }));
            //pass the event up to the parent element who decides what to do with it
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note

            })

        }
    }
    onRemove = (e) => {
        e.preventDefault();
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/dashboard');
    }
    /* Component rendering */
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.submitError ?
                    <p className="form__error">{this.state.submitError}</p>
                    : undefined}
                <input
                    className="text-input"
                    type="text"
                    placeholder="description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    className="text-input"
                    type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    className="text-area"
                    placeholder="Add a note for your expense(optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >

                </textarea>

                <div className="spaced__content">
                    <button className="button">{this.props.mode} Expense</button>
                    {this.props.mode === 'Edit' ?
                        <button className="button button__link__dark" onClick={this.onRemove} type="button">Remove</button> :
                        undefined}
                </div>




            </form>
        );
    }
}



const mapDispatchToProps = (dispatch, props) => ({
    removeExpense: (data) => dispatch(removeExpense(data))
})

const bootstraper = function (state, props) {
    return state;
}
export default connect(bootstraper, mapDispatchToProps)(ExpenseForm);