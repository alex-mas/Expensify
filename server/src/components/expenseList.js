import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './expenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => {
    return (
        <div className="content-container">
            <div className="list-header">
                <div className="show-for-mobile">
                    Expenses
                </div>
                <div className="show-for-desktop">
                    Expense
                </div>
                <div className="show-for-desktop">
                    Amount
                </div>
            </div>
            <div className="list-body">
                {
                    props.expenses.length === 0 ? (
                        <div className="list-item list-item--message">
                            <h1>No expenses</h1>
                        </div>
                    ) : (
                            props.expenses.map((expense) => {
                                return <ExpenseListItem key={expense.id} {...expense} />
                            })
                        )
                }

            </div>


        </div>
    )
};


//connect is sort of a middleware that returns the state and embeds it into the react element provided as argument in the concatenated call
//if we want to pull limited properties away from the state we need to define a function inside the first call
//that returns an object with the properties we want


const bootstraper = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
}
export default connect(bootstraper)(ExpenseList);

