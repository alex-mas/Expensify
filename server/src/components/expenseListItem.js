import React from 'react';
import { Link } from 'react-router-dom'
import moment from 'moment';
import numeral from 'numeral';

//The amount is in cents not dollars thats why we multiply it by 100
const ExpenseListItem = (props) => {
    return (
        <Link className="list-item" to={`/edit/${props.id}`}>
            <div>
                <h3 className="list-item__title">{props.description}</h3>
                <span className="list-item__sub-title">
                    {moment(props.createdAt).format("MMMM Do,  YYYY")}
                </span>
            </div>
            <h3 className="list-item__data">{numeral((props.amount / 100)).format('$0,0.00')}</h3>
        </Link>
    )
};

export default ExpenseListItem;