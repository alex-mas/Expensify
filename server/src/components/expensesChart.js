import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import Chart from 'chart.js';
import moment from 'moment';





export class ExpensesChart extends React.Component {
    state = {
        chart: undefined,
        chartElement: null
    };
    componentDidMount() {
        this.renderChart();
    }
    componentDidUpdate() {
        this.renderChart();
    }
    renderChart() {
        if (this.state.chartElement === null) return;
        if (this.props.expenses.length < 1) {
            this.state.chart = undefined;
            return;
        }
        this.state.chart = new Chart(this.state.chartElement.getContext('2d'), {
            type: 'bar',
            data: {
                labels: this.props.expenses.map((expense) => {
                    return expense.description + ' (' + moment(expense.createdAt).format("dd/M - YY")  + ')';
                }),
                datasets: [{
                    label: 'Expenses',
                    data: this.props.expenses.map((expense) => {
                        return expense.amount;
                    }),
                    backgroundColor: this.props.expenses.map((expense) => {
                        return 'rgb(21, 134, 199)'
                    })
                }]
            },
            options: {
                scales: {
                    yAxes: [{
                        ticks: {
                            beginAtZero: true
                        }
                    }]
                },
                events: [],
                title: {
                    display: true,
                    text: 'Expenses',
                    fontSize: 20
                },
                responsive: true
            }
        });

    }
    render() {
        return (
            <div className="content-container">
                <div className="chartWrapper">
                    {this.props.expenses.length > 0 ?
                        <canvas
                            ref={(elmnt) => {
                                this.state.chartElement = elmnt;

                            }}
                        ></canvas>
                        : undefined
                    }
                </div>
            </div>
        );
    }

}



const bootstraper = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters),
        startDate: state.filters.startDate,
        endDate: state.filters.endDate
    };
}
export default connect(bootstraper)(ExpensesChart);

