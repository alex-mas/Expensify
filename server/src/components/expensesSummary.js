import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expensesTotal';
import { Link } from 'react-router-dom';



export const ExpensesSummary = (props) => {
    const expensesWord = props.expensesCount > 1 ? 'expenses' : 'expense';
    const hiddenDescription = props.hiddenExpensesCount > 1 ? ['are', 'expenses', 'them'] : ['is', 'expense', 'it'];
    return (
        <div className="page-header">
            <div className="content-container">
                {
                    props.expensesCount > 0 ? (
                        <h1 className="page-header__title">Viewing <span>{props.expensesCount}</span> {expensesWord} totaling <span>{numeral((props.expensesTotal / 100)).format('$0,0.00')}</span> ({props.hiddenExpensesCount} hidden)</h1>
                    ) : props.hiddenExpensesCount > 0 ?
                            (<h1 className="page-header__title">There {hiddenDescription[0]} <span>{props.hiddenExpensesCount}</span> hidden {hiddenDescription[1]}, change the filters to see {hiddenDescription[2]}</h1>)
                            : (<h1 className="page-header__title">No expenses found, add some to get the party started!</h1>)
                }
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>

        </div>
    );
};


//bootstraps data from the state into the component properties
const bootstraper = (state) => {
    const expenses = selectExpenses(state.expenses, state.filters);
    const expensesCount = state.expenses.length;
    return {
        hiddenExpensesCount: expensesCount - expenses.length,
        expensesCount: expenses.length,
        expensesTotal: getExpensesTotal(expenses)
    };
}
export default connect(bootstraper)(ExpensesSummary);

