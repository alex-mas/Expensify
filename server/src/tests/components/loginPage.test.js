import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import toJSON from 'enzyme-to-json';
import {shallow} from 'enzyme';
import {LoginPage} from '../../components/login';


describe('Header component unit tests', function(){
    test('should render header correctly',()=>{
        const wrapper = shallow(<LoginPage/>);
        expect(wrapper).toMatchSnapshot();
    });

    test('should call startLogOut on button click',function(){
        const startLogin = jest.fn();
        const wrapper = shallow(<LoginPage startLogin={startLogin}/>);
        wrapper.find('button').simulate('click');
        expect(startLogin).toHaveBeenCalled();
    });
});
