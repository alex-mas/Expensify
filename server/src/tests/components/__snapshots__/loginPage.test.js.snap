// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component unit tests should render header correctly 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      Expensify
    </h1>
    <p>
      put a limit to your expenditure
    </p>
    <button
      className="button"
    >
      Login with Google
    </button>
  </div>
</div>
`;
